### PR TITLE
CA1307: Only consider equal or more accessible overloads

### LINKED ISSUE #6943: `CA1307` does not respect visibility
### Analyzer

**Diagnostic ID**: [CA1307](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1307): `Specify StringComparison for clarity`

### Analyzer source
.NET 8
### Describe the bug

``` csharp
F.M(""); // warning here, even though M(string, StringComparison) is private

class F
{
    private static void M(string s, StringComparison c) { }
    public static void M(string s) { }
}
```

### Expected behavior
no warning

### Actual behavior
warning displayed

### PR DESCRIPTION
**Pull request by david-acker**
Fixes #6943

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs [hunk 1]
```diff
@@ -208,5 +210,35 @@ private static ParameterInfo GetParameterInfo(INamedTypeSymbol type, bool isArra
...
+
+        private static IEnumerable<IMethodSymbol> GetAccessibleMethodsWithSameNameAsTargetMethod(
+            IInvocationOperation invocationExpression,
+            IMethodSymbol targetMethod)
+        {
+            var invocationStart = invocationExpression.Syntax.GetLocation().SourceSpan.Start;
+
+            var methodsWithSameNameAsTargetMethod =
+                invocationExpression.SemanticModel?
+                    .LookupSymbols(
```
Comment (david-acker) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs:222]: I might just be misunderstanding how this method is used, but I noticed this wasn't returning protected methods--hence the extra call to `GetMembers` to get those. Is there a simpler way to handle this?

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs:222]: I would suggest not using LookupSymbols here as it is generally an expensive method call.

Instead I would build on top of existing code and add a `Where` clause to filter out methods whose static-ness doesn't match OR [SemanticModel.IsAccessible](https://sourceroslyn.io/Microsoft.CodeAnalysis/R/bfd11df5ee9625f2.html) returns `false`.

Comment (david-acker) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs]: I've replaced the call to `LookupSymbols` with a `Where` clause that checks `IMethodSymbol.IsStatic` and `SemanticModel.IsAccessible`. 

How should the accessibility check be handled when the `SemanticModel` is null? Currently I have the filter set up to treat the method as if it's not accessible in this situation.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs [hunk 2]
```diff
@@ -208,5 +213,24 @@ private static ParameterInfo GetParameterInfo(INamedTypeSymbol type, bool isArra
         {
             return ParameterInfo.GetParameterInfo(type, isArray, arrayRank, isParams);
         }
+
+        private static IEnumerable<IMethodSymbol> GetAccessibleMethodsWithSameNameAsTargetMethod(
+            IInvocationOperation invocationExpression,
+            IMethodSymbol targetMethod,
+            SemanticModel? semanticModel)
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs:220]: `Operation.SemanticModel` is guaranteed to be non-null for analyzer callbacks. You can remove the null check for semanticModel.
```suggestion
            SemanticModel semanticModel)
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs [hunk 3]
```diff
@@ -102,7 +102,12 @@ protected override void InitializeWorker(CompilationStartAnalysisContext context
                 // plus as additional StringComparison parameter. Default StringComparison may or may not match user's intent,
                 // but it is recommended to explicitly specify it for clarity and readability:
                 // https://learn.microsoft.com/dotnet/standard/base-types/best-practices-strings#recommendations-for-string-usage
-                IEnumerable<IMethodSymbol> methodsWithSameNameAsTargetMethod = targetMethod.ContainingType.GetMembers(targetMethod.Name).OfType<IMethodSymbol>();
+                var methodsWithSameNameAsTargetMethod =
+                    GetAccessibleMethodsWithSameNameAsTargetMethod(
+                        invocationExpression,
+                        targetMethod,
+                        oaContext.Operation.SemanticModel);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs:109]:
```suggestion
                        oaContext.Operation.SemanticModel!);
```

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs:109]: Actually you don't need to pass in `oaContext.Operation.SemanticModel` as a parameter. You can directly access it off  `invocationExpression.SemanticModel`

**Review by mavasani — APPROVED** [review]
Thanks!

**Pull request merged**